typed-function.js

```diff
@@ -965,21 +965,102 @@
+      }
+
+      if (
+        typeof resolvedReference !== 'function' && (
+          isReferTo(resolvedReference) ||
+          isReferToSelf(resolvedReference)
+        )
+      ) {
+        throw new TypeError('Cannot refer to signature "' + signature + '": ' +
+          'signature is referring to a signature "' + reference + '" which is not yet resolved');
+      }
+    }
+
+    function resolveReferences(rawSignaturesMap, self) {
+      // we must not make changes in the original signatures map
+      var signaturesMap = shallowClone(rawSignaturesMap);
+
+      // resolve references, providing "self"
+      for (var signature in signaturesMap) {
+        if (signaturesMap.hasOwnProperty(signature)) {
+          var fn = signaturesMap[signature];
+
+          if (isReferTo(fn)) {
+            var resolvedReferences = fn.referTo.references.map(reference => {
+              var resolvedReference = signaturesMap[reference];
+
+              verifyResolvedReference(resolvedReference, reference, signature);
+
+              return resolvedReference;
+            })
+            var resolved = fn.referTo.callback.apply(this, resolvedReferences);
+
+            // we need to keep the .referTo object (with references and callback)
+            // to be able to merge typed-functions later on: then we need to
+            // resolve the new signatures again.
+            resolved.referTo = fn.referTo;
+
+            signaturesMap[signature] = resolved;
+          }
+
+          if (isReferToSelf(fn)) {
+            var resolvedSelf = fn.referToSelf.callback(self);
+
+            // we need to keep the .referTo object (with references and callback)
+            // to be able to merge typed-functions later on: then we need to
+            // resolve the new signatures again.
+            resolvedSelf.referToSelf = fn.referToSelf;
+
+            signaturesMap[signature] = resolvedSelf;
+          }
+        }
+      }
+
+      return signaturesMap;
+    }
+
     /**
      * Create a typed function
      * @param {String} name               The name for the typed function
-     * @param {Object.<string, function>} signaturesMap
+     * @param {Object.<string, function>} rawSignaturesMap
      *                                    An object with one or
      *                                    multiple signatures as key, and the
      *                                    function corresponding to the
      *                                    signature as value.
      * @return {function}  Returns the created typed function.
      */
-    function createTypedFunction(name, signaturesMap) {
-      if (Object.keys(signaturesMap).length === 0) {
+    function createTypedFunction(name, rawSignaturesMap) {
+      if (Object.keys(rawSignaturesMap).length === 0) {
         throw new SyntaxError('No signatures provided');
       }
 
+      // resolve references, providing "self"
+      var signaturesMap = resolveReferences(rawSignaturesMap, fn);
+
```

Comment:
This is actually critical, so I think not just a recommendation, but a necessity. If you resolve references before splitting the signatures, then you would have to write the above example as
```
const refer = typed({
  'number|string': arg => 'wow: ' + arg,
  'boolean': typed.referTo('number|string', refS => b => {
    if (b) return refS('true that!')
    return refS('no way...')
  })
})
```
But now this function is unmergeable:
```
const reref = typf(refer, typed({Array: a => 'Got ' + a.length + 'items'}))
```
results in
```
TypeError: Cannot refer to signature "boolean": reference signature "number|string" not found
```
because of course once it's fully constructed, `refer` has no signature "number|string", it has two signatures "number" and "string" that correspond to the same implementation (`arg => 'wow' + arg`).
[By the way, it was trying to understand this example that led to my filing #141 and the associated PR #142.]

Reply:
Very good point! Will fix that.

Reply:
P.S. if typed-function did allow forward references, one wouldn't have to worry about the relative order of sorting the signatures and resolving references, as is currently carefully commented in the code.

Reply:
Yes, let's see if we can directly make that working! I'll close this conversation since the code works correctly now for this case, but the refactoring still has to be addressed.